### ISSUE TITLE
Caching should be disabled in tests (or at least not use HOME cache)

### ISSUE THREAD

**Issue by ocharles** [dhall-lang/dhall-haskell#1192]
I just had a really weird problem, where a test was saying that an expression type checked when it should have failed, but whenever I loaded the expression into Dhall myself it failed to type check. I eventually narrowed it down to this:

```
[nix-shell:~/work/dhall-haskell/dhall]$ cabal new-run exe:dhall -- resolve <<< './dhall-lang/tests/typecheck/failure/unit/HeterogenousToMap.dhall'
Up to date
[ { mapKey = "bar", mapValue = "Bar" }, { mapKey = "foo", mapValue = 1 } ]

[nix-shell:~/work/dhall-haskell/dhall]$ cabal new-run exe:dhall -- type < './dhall-lang/tests/typecheck/failure/unit/HeterogenousToMap.dhall'
Up to date

Use "dhall --explain" for detailed errors

Error: ❰toMap❱ expects a homogenous record

1│ toMap { foo= 1, bar= "Bar" }

(stdin):1:1
```

In the process of working on a branch, I must have written some code that caused the cache to get poisoned. `dhall resolve` was never type checking it again, just returning whatever the cache had. Moderately concerning, but no biggy. The bigger problem was the test suite is actually hitting that cache too.

TL;DR: I don't think `~/.cache` should be accessible in tests.

**Comment by Gabriel439** [comment]
@ocharles: The tests don't use `~/.cache`.  See: https://github.com/dhall-lang/dhall-haskell/blob/master/dhall/tests/Dhall/Test/Main.hs#L70.  What happened to you is that the cache for your test run was poisoned by a prior test run.

**Comment by ocharles** [comment]
Hmm, I'm confused. I literally got the test to pass by removing ~/.cache, though I did also remove all my dist directories - but I don't think that's what PWD would have been. Will look more next week

**Comment by sjakobi** [comment]
I think #1275 somewhat improves the situation by ignoring the semantic cache for the type-inference tests:

https://github.com/dhall-lang/dhall-haskell/pull/1275/files#diff-412cfdcdac2676c6a5740d7a459814aeR63-R66

To prevent cache poisoning between test runs it might make sense to create the cache directory in a fresh temporary directory.

**Comment by Gabriel439** [comment]
I think this was fixed recently by https://github.com/dhall-lang/dhall-haskell/pull/2185
